RulesEngine/OATScopedRegexOperation.cs

```diff
@@ -0,0 +1,109 @@
+﻿using Microsoft.CST.OAT;
+using Microsoft.CST.OAT.Operations;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.ApplicationInspector.RulesEngine
+{
+    public class OATScopedRegexOperation : OatOperation
+    {
+        public OATScopedRegexOperation(Analyzer analyzer) : base(Operation.Custom, analyzer)
+        {
+            CustomOperation = "ScopedRegex";
+            regexEngine = new RegexOperation(analyzer);
+            OperationDelegate = ScopedRegexOperationDelegate;
+            ValidationDelegate = ScopedRegexValidationDelegate;
+        }
+
+        public OperationResult ScopedRegexOperationDelegate(Clause c, object? state1, object? state2, IEnumerable<ClauseCapture>? captures)
+        {
+            if (state1 is TextContainer tc && c is OATScopedRegexClause src)
+            {
+                var regexOpts = RegexOptions.Compiled;
+                if (src.Arguments.Contains("i"))
+                {
+                    regexOpts |= RegexOptions.IgnoreCase;
+                }
+                if (src.Arguments.Contains("m"))
+                {
+                    regexOpts |= RegexOptions.Multiline;
+                }
+                var boundaries = new List<Boundary>();
+                var target = tc.Target;
+                if (Analyzer != null)
+                {
+                    foreach (var pattern in src.Data.Select(x => regexEngine.StringToRegex(x, regexOpts)))
+                    {
+                        if (pattern is Regex r)
+                        {
+                            var matches = r.Matches(target);
+                            foreach (var match in matches)
+                            {
+                                if (match is Match m)
+                                {
+                                    Boundary translatedBoundary = new Boundary()
+                                    {
+                                        Length = m.Length,
+                                        Index = m.Index + tc.GetLineBoundary(tc.LineNumber).Index
+                                    };
+                                    // Should return only scoped matches
+                                    if (tc.ScopeMatch(src.Scopes, translatedBoundary))
+                                    {
+                                        boundaries.Add(translatedBoundary);
+                                    }
+                                }
+                            }
+                        }
+
+                        var result = src.Invert ? boundaries.Count == 0 : boundaries.Count > 0;
+                        return new OperationResult(result, result && src.Capture ? new TypedClauseCapture<List<Boundary>>(c, boundaries, state1) : null);
```

Comment:
This returns inside the foreach loop so it will always only use the first pattern.

Reply:
This was a bug in DevSkim as well that I've now fixed.

Reply:
cool I'll grab from there